File: app/src/androidTest/java/io/github/aksulightning/flyby/SharedTestProvider.java
package io.github.aksulightning.flyby;

import android.database.Cursor;
import android.database.MatrixCursor;
import android.os.CancellationSignal;
import android.os.ParcelFileDescriptor;
import android.provider.DocumentsContract.Document;
import android.provider.DocumentsProvider;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

/** SDK-only fixture: its separate test-APK process cannot rely on the target's Kotlin runtime. */
public class SharedTestProvider extends DocumentsProvider {
    private File root() {
        File root = new File(getContext().getFilesDir(), "shared-test");
        root.mkdirs();
        return root;
    }
    private File file(String id) throws FileNotFoundException {
        if (!id.equals("root") && !id.startsWith("root/")) throw new FileNotFoundException();
        try {
            File root = root().getCanonicalFile();
            File f = id.equals("root") ? root : new File(root, id.substring(5)).getCanonicalFile();
            if (!f.toPath().startsWith(root.toPath())) throw new FileNotFoundException();
            return f;
        } catch (IOException e) { throw new FileNotFoundException(); }
    }
    private String id(File f) { return f.equals(root()) ? "root" : "root/" + root().toPath().relativize(f.toPath()); }
    @Override public boolean onCreate() { return true; }
    @Override public Cursor queryRoots(String[] projection) { return new MatrixCursor(projection == null ? new String[]{"root_id"} : projection); }
    private Cursor rows(File[] files, String[] projection) {
        String[] columns = projection == null ? new String[]{Document.COLUMN_DOCUMENT_ID, Document.COLUMN_DISPLAY_NAME, Document.COLUMN_MIME_TYPE,
            Document.COLUMN_SIZE, Document.COLUMN_LAST_MODIFIED, Document.COLUMN_FLAGS} : projection;
        MatrixCursor result = new MatrixCursor(columns);
        if (files != null) for (File f : files) {
            Object[] values = new Object[columns.length];
            for (int i = 0; i < columns.length; i++) {
                switch (columns[i]) {
                    case Document.COLUMN_DOCUMENT_ID: values[i] = id(f); break;
                    case Document.COLUMN_DISPLAY_NAME: values[i] = f.getName(); break;
                    case Document.COLUMN_MIME_TYPE: values[i] = f.isDirectory() ? Document.MIME_TYPE_DIR : "application/octet-stream"; break;
                    case Document.COLUMN_SIZE: values[i] = f.length(); break;
                    case Document.COLUMN_LAST_MODIFIED: values[i] = f.lastModified(); break;
                    case Document.COLUMN_FLAGS: values[i] = Document.FLAG_SUPPORTS_WRITE | Document.FLAG_SUPPORTS_DELETE | Document.FLAG_SUPPORTS_RENAME |
                        (f.isDirectory() ? Document.FLAG_DIR_SUPPORTS_CREATE : 0); break;
                }
            }
            result.addRow(values);
        }
        return result;
    }
    @Override public Cursor queryDocument(String documentId, String[] projection) throws FileNotFoundException {
        File f = file(documentId); if (!f.exists()) throw new FileNotFoundException();
        return rows(new File[]{f}, projection);
    }
    @Override public Cursor queryChildDocuments(String parentDocumentId, String[] projection, String sortOrder) throws FileNotFoundException {
        return rows(file(parentDocumentId).listFiles(), projection);
    }
    @Override public ParcelFileDescriptor openDocument(String documentId, String mode, CancellationSignal signal) throws FileNotFoundException {
        return ParcelFileDescriptor.open(file(documentId), ParcelFileDescriptor.parseMode(mode));
    }
    private void name(String name) {
        if (name.isEmpty() || name.equals(".") || name.equals("..") || name.contains("/")) throw new IllegalArgumentException();
    }
    @Override public String createDocument(String parentDocumentId, String mimeType, String displayName) throws FileNotFoundException {
        name(displayName); File f = new File(file(parentDocumentId), displayName);
        try {
            if (!(mimeType.equals(Document.MIME_TYPE_DIR) ? f.mkdir() : f.createNewFile())) throw new FileNotFoundException();
        } catch (IOException e) { throw new FileNotFoundException(); }
        return id(f);
    }
    @Override public void deleteDocument(String documentId) throws FileNotFoundException {
        if (documentId.equals("root") || !file(documentId).delete()) throw new FileNotFoundException();
    }
    @Override public String renameDocument(String documentId, String displayName) throws FileNotFoundException {
        name(displayName); if (documentId.equals("root")) throw new FileNotFoundException();
        File f = file(documentId), next = new File(f.getParentFile(), displayName);
        if (!f.renameTo(next)) throw new FileNotFoundException();
        return id(next);
    }
    @Override public boolean isChildDocument(String parentDocumentId, String documentId) {
        try { return file(documentId).toPath().startsWith(file(parentDocumentId).toPath()); }
        catch (FileNotFoundException e) { return false; }
    }
}

File: app/src/androidTest/java/io/github/aksulightning/flyby/SharedTestProvider.kt
package io.github.aksulightning.flyby

import android.database.Cursor
import android.database.MatrixCursor
import android.os.CancellationSignal
import android.os.ParcelFileDescriptor
import android.provider.DocumentsContract.Document
import android.provider.DocumentsProvider
import java.io.File

/** Disposable document provider installed only with the instrumentation APK. */
class SharedTestProvider : DocumentsProvider() {
    private val root get() = File(requireNotNull(context).filesDir, "shared-test").apply { mkdirs() }.canonicalFile
    private fun file(id: String): File {
        val f = if (id == "root") root else File(root, id.removePrefix("root/")).canonicalFile
        require(id == "root" || id.startsWith("root/"))
        require(f.toPath().startsWith(root.toPath()))
        return f
    }
    private fun id(f: File) = if (f == root) "root" else "root/" + f.relativeTo(root).path
    override fun onCreate() = true
    override fun queryRoots(projection: Array<out String>?) = MatrixCursor(projection ?: arrayOf("root_id"))
    private fun rows(files: List<File>, projection: Array<out String>?): Cursor {
        val columns = projection ?: arrayOf(Document.COLUMN_DOCUMENT_ID, Document.COLUMN_DISPLAY_NAME, Document.COLUMN_MIME_TYPE, Document.COLUMN_SIZE, Document.COLUMN_LAST_MODIFIED, Document.COLUMN_FLAGS)
        return MatrixCursor(columns).apply { files.forEach { f -> addRow(columns.map { column -> when (column) {
            Document.COLUMN_DOCUMENT_ID -> id(f)
            Document.COLUMN_DISPLAY_NAME -> f.name
            Document.COLUMN_MIME_TYPE -> if (f.isDirectory) Document.MIME_TYPE_DIR else "application/octet-stream"
            Document.COLUMN_SIZE -> f.length()
            Document.COLUMN_LAST_MODIFIED -> f.lastModified()
            Document.COLUMN_FLAGS -> Document.FLAG_SUPPORTS_WRITE or Document.FLAG_SUPPORTS_DELETE or Document.FLAG_SUPPORTS_RENAME or (if (f.isDirectory) Document.FLAG_DIR_SUPPORTS_CREATE else 0)
            else -> null
        } }.toTypedArray()) } }
    }
    override fun queryDocument(documentId: String, projection: Array<out String>?) = rows(listOf(file(documentId)), projection)
    override fun queryChildDocuments(parentDocumentId: String, projection: Array<out String>?, sortOrder: String?) = rows(file(parentDocumentId).listFiles()?.toList().orEmpty(), projection)
    override fun openDocument(documentId: String, mode: String, signal: CancellationSignal?) = ParcelFileDescriptor.open(file(documentId), ParcelFileDescriptor.parseMode(mode))
    override fun createDocument(parentDocumentId: String, mimeType: String, displayName: String): String {
        require(displayName != "." && displayName != ".." && '/' !in displayName)
        val f = File(file(parentDocumentId), displayName)
        check(if (mimeType == Document.MIME_TYPE_DIR) f.mkdir() else f.createNewFile())
        return id(f)
    }
    override fun deleteDocument(documentId: String) { require(documentId != "root"); check(file(documentId).delete()) }
    override fun renameDocument(documentId: String, displayName: String): String {
        require(documentId != "root" && displayName != "." && displayName != ".." && '/' !in displayName)
        val f = file(documentId); val next = File(f.parentFile, displayName)
        check(f.renameTo(next)); return id(next)
    }
    override fun isChildDocument(parentDocumentId: String, documentId: String) = file(documentId).toPath().startsWith(file(parentDocumentId).toPath())
}
